### PR TITLE
HTTP2: fix race between request failure and response body read

### LINKED ISSUE #39729: PostAsyncDuplex_RequestContentExceptionAfterResponseEndReceivedButBeforeConsumed_ResetsStreamAndThrowsOnResponseStreamRead failed in CI
``` System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2/PostAsyncDuplex_RequestContentExceptionAfterResponseEndReceivedButBeforeConsumed_ResetsStreamAndThrowsOnResponseStreamRead

Windows.7.Amd64.Open-x86-Release
Get Repro environment
Unhandled Exception of Type Xunit.Sdk.ThrowsException
Message :
Assert.Throws() Failure
Expected: typeof(System.IO.IOException)
Actual:   (No exception was thrown)
Stack Trace :
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.PostAsyncDuplex_RequestContentExceptionAfterResponseEndReceivedButBeforeConsumed_ResetsStreamAndThrowsOnResponseStreamRead() in /_/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:line 2252
--- End of stack trace from previous location where exception was thrown ---
```

### PR DESCRIPTION
**Pull request by geoffkizer**
Fixes #39729

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 1]
```diff
@@ -169,6 +169,10 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                 {
                     if (NetEventSource.IsEnabled) Trace($"Failed to send request body: {e}");
 
+                    // Cancel the stream before we set _requestCompletionState below.
+                    // Otherwise, a response stream reader may race with the actual Cancel.
+                    Cancel();
+
                     lock (SyncObject)
                     {
                         Debug.Assert(_requestCompletionState == StreamCompletionState.InProgress, $"Request already completed with state={_requestCompletionState}");
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Is this assert still valid?  The Cancel couldn't cause the state to change?

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: It's still valid. Cancel will not modify _requestCompletionState.

**Comment by geoffkizer** [comment]
@dotnet/dnceng I'm getting "500: Internal Server Error" trying to look at test failures. Been like this for an hour or so.

**Comment by MattGal** [comment]
@geoffkizer  That'd be  https://github.com/dotnet/core-eng/issues/7124, I'm rolling out a hotfix for it now.

**Comment by danmosemsft** [comment]
Can we please get stress done on this before porting to 3.0

cc @eiriktsarpalis

**Comment by geoffkizer** [comment]
I'm running stress on it now -- 2 minutes in, so far so good :)

**Comment by geoffkizer** [comment]
HttpClient functional failures are unrelated (CurlHandler)

**Comment by geoffkizer** [comment]
Ran stress against http.sys for about an hour and a half, 0 failures

**Comment by karelz** [comment]
This PR went into master - fixing milestone to 5.0.
It was later ported into release/3.0 branch in PR #39781.

**Pull request merged**